syntax = "proto3";

package gtools.state;

import "growtopia.proto";

enum StateUpdateWhat {
  STATE_UNSPECIFIED = 0;
  STATE_PLAYER_UPDATE = 1;
  STATE_SET_MY_PLAYER = 2;
  STATE_SEND_INVENTORY = 3;
  STATE_MODIFY_INVENTORY = 4;
  STATE_ENTER_WORLD = 5;
  STATE_EXIT_WORLD = 6;
  STATE_PLAYER_JOIN = 8;
  STATE_PLAYER_LEAVE = 9;
  STATE_MODIFY_WORLD = 10;
  STATE_MODIFY_WORLD_BATCHED = 15;
  STATE_MODIFY_ITEM = 11;
  STATE_UPDATE_STATUS = 12;
  STATE_SET_CHARACTER_STATE = 13;
  STATE_SET_MY_TELEMETRY = 14;
  STATE_SEND_LOCK = 16;
  STATE_UPDATE_TREE_STATE = 17;
  STATE_TILE_CHANGE_REQUEST = 18;
  STATE_NPC_UPDATE = 19;
}

message StateUpdate {
  StateUpdateWhat what = 1;
  oneof update {
    PlayerUpdate player_update = 2;
    uint32 set_my_player = 3;
    gtools.growtopia.Inventory send_inventory = 4;
    ModifyInventory modify_inventory = 5;
    EnterWorld enter_world = 6;
    gtools.growtopia.Player player_join = 7;
    uint32 player_leave = 8;
    ModifyWorld modify_world = 13;
    ModifyWorldBatched modify_world_batched = 14;
    ModifyItem modify_item = 9;
    uint32 update_status = 10;
    gtools.growtopia.CharacterState character_state = 11;
    SetMyTelemetry set_my_telemetry = 12;
    SendLock send_lock = 15;
    UpdateTreeState update_tree_state = 16;
    TileChangeRequest tile_change_req = 17;
    NpcUpdate npc_update = 18;
  }
}

message NpcRemove {
  uint32 id = 1;
  uint32 state = 2;
}

message NpcUpdatePos {
  uint32 id = 6;
  int32 param1 = 1;  // 0 means ignore for param
  int32 param2 = 2;
  float param3 = 3;
  float x = 4;
  float y = 5;
}

message NpcUpdate {
  enum Op {
    OP_UNSPECIFIED = 0;
    OP_ADD = 1;
    OP_REMOVE = 3;
    OP_RESET_STATE = 2;
    OP_UPDATE_TARGET = 4;
    OP_UPDATE_POS = 5;
  }

  Op op = 1;
  oneof payload {
    gtools.growtopia.Npc npc = 2;
    uint32 id = 3;
    NpcRemove remove = 4;
    NpcUpdatePos update_target = 5;
  }
}

message TileChangeRequest {
  int32 x = 1;
  int32 y = 2;
  uint32 id = 3;
  uint32 flags = 4;
  bool splice = 6;
  uint32 seed_id = 5;
}

message UpdateTreeState {
  int32 x = 1;
  int32 y = 2;
  uint32 item_id = 5;
  bool harvest = 6;
  bool add_spawn_seeds_flag = 3;
  bool add_seedling_flag = 4;
}

message SendLock {
  int32 x = 1;
  int32 y = 2;
  int32 lock_owner_id = 3;
  uint32 lock_item_id = 4;
  repeated uint32 tiles_affected = 5;
}

message ModifyWorld {
  enum Op {
    OP_UNSPECIFIED = 0;
    OP_REPLACE = 1;
    OP_PLACE = 2;
    OP_DESTROY = 3;
    OP_UPDATE_EXTRA_DATA = 4;
  }
  Op op = 1;
  oneof payload {
    gtools.growtopia.Tile tile = 2;
    bytes extra = 3;
  }
}

message ModifyWorldBatched { repeated ModifyWorld events = 1; }

message SetMyTelemetry {
  uint32 server_ping = 5;
  uint32 client_ping = 8;
  float time_since_login = 6;
  float time_in_world = 7;
}

message ModifyItem {
  enum Op {
    OP_UNSPECIFIED = 0;
    OP_CREATE = 1;
    OP_SET_AMOUNT = 3;
    OP_TAKE = 4;
  }

  Op op = 1;
  uint32 item_id = 2;
  uint32 uid = 3;
  uint32 amount = 6;
  float x = 4;
  float y = 5;
  uint32 flags = 7;
}

message EnterWorld {
  gtools.growtopia.World enter_world = 1;
  bytes door_id = 2;
}

message ModifyInventory {
  uint32 id = 1;
  int32 to_add = 2;
}

message PlayerUpdate {
  int32 net_id = 1;
  float x = 2;
  float y = 3;
  uint32 flags = 4;
}
